pinot-spi/src/main/java/org/apache/pinot/spi/stream/RowMetadata.java

```diff
@@ -64,24 +62,34 @@ default long getFirstStreamRecordIngestionTimeMs() {
   }
 
   /**
-   * Returns the stream message headers
-   *
-   * @return A {@link GenericRow} that encapsulates the headers in the ingested row
+   * Returns the stream offset of the message.
+   */
+  @Nullable
```

Comment:
out of curiosity, why do we use nulls and nullable annnotation instead of Optionals?

Reply:
Both historic reason and performance reason:
- Optional was introduced in java 8 (it is handy to be used along with function APIs), and majority of the code is using `null`
- There could be performance overhead because it is an extra Object wrapper (24 bytes) over a reference. In majority of the places it doesn't matter, but checkout this interesting [post](https://pkolaczk.github.io/overhead-of-optional/)

Reply:
that's a solid post. thank you for the reference!